Fix module loading in logging config

viiccwen opened the pull request:
## Summary

Fix issue where custom logging configuration with simple module paths (like `log_config.LOGGING_CONFIG`) caused triggerer and scheduler to fail during startup.

## Problem

When users set `logging_config_class = log_config.LOGGING_CONFIG` in their `airflow.cfg`, the `load_logging_config()` function would fail because:

1. `logging_class_path.rsplit(".", 1)[0]` produces `"log_config"`
2. `import_string("log_config")` fails because it expects a dot-separated path like `"module.submodule.attribute"`
3. This caused `ImportError` and prevented `REMOTE_TASK_LOG` and `DEFAULT_REMOTE_CONN_ID` from being loaded
4. In some environments, this prevented triggerer and scheduler from starting

## Solution

Replace `import_string()` with `importlib.import_module()` in `load_logging_config()`

## Changes

### Code
- in`airflow-core/src/airflow/logging_config.py`, with`load_logging_config()`

### Test
- in `airflow-core/tests/unit/core/test_logging_config.py`
- **Added**: Parametrized test for different module path structures
- **Added**: Test for graceful fallback when remote logging vars are missing
- **Added**: Helper method to reduce test code duplication

## Related Issues

resolves #54414

viiccwen committed:
fix(logging): fix module loading by using `import_module()`

- Replace `import_string()` with `import_module()` in `load_logging_config()`
- Fix issue where simple module paths like `log_config.LOGGING_CONFIG` failed
- Add `importlib` import inside function to avoid circular imports

viiccwen committed:
fix(logging): fix module loading by using `import_module()`

- Replace `import_string()` with `import_module()` in `load_logging_config()`
- Fix issue where simple module paths like `log_config.LOGGING_CONFIG` failed
- Add `importlib` import inside function to avoid circular imports

viiccwen committed:
fix(logging): fix module loading by using `import_module()`

- Replace `import_string()` with `import_module()` in `load_logging_config()`
- Fix issue where simple module paths like `log_config.LOGGING_CONFIG` failed
- Add `importlib` import inside function to avoid circular imports

viiccwen committed:
test(logging): add tests for logging config module path handling

- Add parametrized test for simple and nested module paths
- Add test for graceful fallback when remote logging vars are missing
- Create helper method to reduce test code duplication
- Test both log_config.LOGGING_CONFIG and nested.module.LOGGING_CONFIG scenarios

Ensures logging config works with various module path structures

viiccwen committed:
test(logging): add tests for logging config module path handling

- Add parametrized test for simple and nested module paths
- Add test for graceful fallback when remote logging vars are missing
- Create helper method to reduce test code duplication
- Test both log_config.LOGGING_CONFIG and nested.module.LOGGING_CONFIG scenarios

Ensures logging config works with various module path structures

viiccwen committed:
test(logging): add tests for logging config module path handling

- Add parametrized test for simple and nested module paths
- Add test for graceful fallback when remote logging vars are missing
- Create helper method to reduce test code duplication
- Test both log_config.LOGGING_CONFIG and nested.module.LOGGING_CONFIG scenarios

Ensures logging config works with various module path structures

jason810496 commented on airflow-core/src/airflow/logging_config.py:
```diff
@@ -73,7 +73,10 @@ def load_logging_config() -> tuple[dict[str, Any], str]:
         try:
-            mod = import_string(modpath)
+            # Import here to avoid circular imports
+            from importlib import import_module
```
IMO, it's fine to move the import of `importlib` to top level, as `airflow.utils.module_loading` will also import `importlib` at top level.

```suggestion
```

prdai replied:
+1

viiccwen replied:
Thank @jason810496 and @Programmer-RD-AI to review it!
I've revised it already.

jason810496 commented on airflow-core/tests/unit/core/test_logging_config.py:
```diff
@@ -365,3 +474,40 @@ def test_loading_remote_logging_with_hdfs_handler(self):
+                logging_config, _ = load_logging_config()
+
+                assert isinstance(logging_config, dict)
+                assert logging_config["version"] == 1
```
Should we leverage `self._verify_basic_logging_config` utility as well?

viiccwen replied:
Yes, I think so. 
I've revised it already.

jason810496 commented on airflow-core/tests/unit/core/test_logging_config.py:
```diff
@@ -365,3 +474,40 @@ def test_loading_remote_logging_with_hdfs_handler(self):
+            ),
+        ],
+    )
+    def test_load_logging_config_module_paths(self, settings_content, module_structure, expected_path):
```
```suggestion
    def test_load_logging_config_module_paths(self, settings_content: str, module_structure: str, expected_path: str):
```

viiccwen replied:
thx!

jason810496 commented on airflow-core/tests/unit/core/test_logging_config.py:
```diff
@@ -191,6 +293,13 @@ def teardown_method(self):
         importlib.reload(airflow_local_settings)
         configure_logging()
 
+    def _verify_basic_logging_config(self, logging_config, logging_class_path, expected_path):
```
```suggestion
    def _verify_basic_logging_config(self, logging_config: dict, logging_class_path: str, expected_path: str):
```

viiccwen replied:
thx!

jason810496 commented on airflow-core/tests/unit/core/test_logging_config.py:
```diff
@@ -96,6 +96,108 @@
+DEFAULT_REMOTE_CONN_ID = "test_conn_id"
+"""
+
+SETTINGS_FILE_NESTED_MODULE = """
```
It seems that the content of `SETTINGS_FILE_NESTED_MODULE ` is exactly same as `SETTINGS_FILE_SIMPLE_MODULE` expect the last line ( `DEFAULT_REMOTE_CONN_ID = "nested_conn_id"` ).
How about we copy from ‎`SETTINGS_FILE_SIMPLE_MODULE` and then replace the last line?

viiccwen replied:
thx! I've revised it below
```python
SETTINGS_FILE_NESTED_MODULE = SETTINGS_FILE_SIMPLE_MODULE.replace(
    'DEFAULT_REMOTE_CONN_ID = "test_conn_id"', 'DEFAULT_REMOTE_CONN_ID = "nested_conn_id"'
)
```

jason810496 reviewed (commented):
Nice! Thanks for the PR! LGTM overall.

viiccwen committed:
Merge remote-tracking branch 'origin/fix-module-loading' into fix-module-loading

viiccwen committed:
Merge remote-tracking branch 'origin/fix-module-loading' into fix-module-loading

viiccwen committed:
refactor(logging): move `import_module` to the top level

viiccwen committed:
test(logging): optimize test constants to reduce code duplication

Replace `SETTINGS_FILE_NESTED_MODULE` replicate part with string replacement from `SETTINGS_FILE_SIMPLE_MODULE`

viiccwen committed:
test(logging): use shared helper method for logging config validation

Replace manual assertions with `self._verify_basic_logging_config` in fallback test to ensure consistent validation across all logging config tests.

jason810496 commented on airflow-core/src/airflow/logging_config.py:
```diff
@@ -24,7 +24,7 @@
 from airflow.configuration import conf
 from airflow.exceptions import AirflowConfigException
-from airflow.utils.module_loading import import_string
+from airflow.utils.module_loading import import_module, import_string
```
It would be better to import the `import_module` from `importlib`, as we should import the function from where it is originally defined.

viiccwen replied:
got it! 😆 thx for correction!

jason810496 commented on airflow-core/tests/unit/core/test_logging_config.py:
```diff
@@ -96,6 +96,77 @@
+    'DEFAULT_REMOTE_CONN_ID = "test_conn_id"', 'DEFAULT_REMOTE_CONN_ID = "nested_conn_id"'
+)
+
+SETTINGS_FILE_NO_REMOTE_VARS = """
```
Sorry, I didn't notice that we could also copy from `SETTINGS_FILE_SIMPLE_MODULE` variable for `SETTINGS_FILE_NO_REMOTE_VARS` and replace the last few lines.

viiccwen replied:
got it! I refactored all `SETTINGS_FILE` with renaming and replacing 🏃

jason810496 reviewed (commented):
Thanks for the fix!

viiccwen committed:
refactor(logging_config): simplify, improve test config vars

- Remove redundant `SETTINGS_FILE_NESTED_MODULE`
- Rename `SETTINGS_FILE_SIMPLE_MODULE` to `SETTINGS_FILE_WITH_REMOTE_VARS`

Lee-W commented on airflow-core/tests/unit/core/test_logging_config.py:
```diff
@@ -96,6 +96,18 @@
+REMOTE_TASK_LOG = None
+DEFAULT_REMOTE_CONN_ID = "test_conn_id"
+"""
+)
```
```suggestion
SETTINGS_FILE_WITH_REMOTE_VARS = (
    f"""{SETTINGS_FILE_VALID}
REMOTE_TASK_LOG = None
DEFAULT_REMOTE_CONN_ID = "test_conn_id"
"""
)
```

Lee-W commented on airflow-core/tests/unit/core/test_logging_config.py:
```diff
@@ -96,6 +96,18 @@
+"""
+)
+
+SETTINGS_FILE_NO_REMOTE_VARS = SETTINGS_FILE_WITH_REMOTE_VARS.replace("REMOTE_TASK_LOG = None", "").replace(
```
Are `SETTINGS_FILE_NO_REMOTE_VARS` and `SETTINGS_FILE_VALID` the same? Do we need another variable for it?

viiccwen replied:
Yes, functionally they are the same. 🤔
I've removed it! thx for pointing out!

Lee-W commented on airflow-core/tests/unit/core/test_logging_config.py:
```diff
@@ -191,6 +203,13 @@ def teardown_method(self):
         importlib.reload(airflow_local_settings)
         configure_logging()
 
+    def _verify_basic_logging_config(self, logging_config: dict, logging_class_path: str, expected_path: str):
```
```suggestion
    def _verify_basic_logging_config(self, logging_config: dict, logging_class_path: str, expected_path: str) -> None:
```

viiccwen replied:
thx!

Lee-W commented on airflow-core/tests/unit/core/test_logging_config.py:
```diff
@@ -365,3 +384,43 @@ def test_loading_remote_logging_with_hdfs_handler(self):
+                mock_log.info.assert_called_with(
+                    "Remote task logs will not be available due to an error:  %s",
+                    mock_log.info.call_args[0][1],
+                )
```
```suggestion
                mock_log.info.mock_calls = [
                    call(
                        "Remote task logs will not be available due to an error:  %s",
                    )
                ]
```

viiccwen replied:
thx for pointing out! 🙏

viiccwen committed:
refactor(logging): imporve & simplify logging testing file

1. remove `SETTING_FILE_NO_REMOTE_VARS` bcz redundant var
2. add return typo with `_verify_basic_logging_config`
3. add unittest.mock `call`
